internal/agent/hostcache.go

```diff
@@ -79,18 +94,28 @@ func SetCache(_ http.ResponseWriter, r *http.Request) {
 	if err := decoder.Decode(&newCache); err != nil {
 		fmt.Println("Error processing cache update")
 	}
-	hostCache = newCache
+	hostCache.Lock()
+	defer hostCache.Unlock()
+	hostCache.internal = newCache
 	fmt.Println("cache was set :)")
 }
 
 // GetCache returns entire cache
 func GetCache(w http.ResponseWriter, _ *http.Request) {
-	if err := json.NewEncoder(w).Encode(hostCache); err != nil {
+	hostCache.RLock()
+	defer hostCache.RUnlock()
+	mapCopy := make(map[string]hosts.Host)
```

Comment:
Why do we need a deep copy of the map?

Reply:
when you return the map the lock is released but maps/slices are returned by reference so when the json encoder reads the map that is returned while writes are happening, you still get a concurrent read write panic.

Reply:
This is fascinating. Sweet work @gbraxton